### ISSUE TITLE
The special exception allowing assigning a self-reference to a reference annotated late is unsafe

### ISSUE THREAD

**Issue by QrczakMK** [ceylon/ceylon-spec#1167]
It allows to observe a partially initialized object instead of detecting this at runtime. Example:

```ceylon
shared void play() {
	class C() {
		late C self;
		self = this;
		shared String x = self.x + "a";
	}
	C c = C();
	print(c.x);
}
```
Prints “nulla”.

**Comment by gavinking** [comment]
So the only really clear solution to this that I can see would be to add a boolean `$isFullyInitialized$` instance variable to every class which assigns a self reference to a `late` attribute in its initializer, and check that instance variable in the getter for the `late` attribute, to ensure that it is not returning a reference to a not-full-initialized class. (Setting the `$isFullyInitialized$` is perhaps slightly tricky once you take into account inheritance, but it should be doable.)

Indeed, I think that's what I originally intended for this functionality; I think we just never actually implemented it.

If that's the solution, then this is a problem for the backends, not for the typechecker.

Anyway, nice catch, @QrczakMK.

**Comment by gavinking** [comment]
> So the only really clear solution to this that I can see would be to add a boolean `$isFullyInitialized$` instance variable

Actually there's a better way, i guess. We can make `$isFullyInitialized$()` a method that checks that no non-null fields have null values. That should be quite easy to implement.

**Comment by gavinking** [comment]
Yeah, that's the right solution, I believe. So I'm closing this issue, after opening ceylon/ceylon-compiler#2031 and ceylon/ceylon-js#496.

**Comment by gavinking** [comment]
Well, perhaps closing the issue is a touch premature, since really this should be nailed down by the spec...

**Comment by gavinking** [comment]
> Actually there's a better way, i guess. We can make `$isFullyInitialized$()` a method that checks that no non-null fields have null values. That should be quite easy to implement. It could even be `$assertFullyInitialized()`.

Actually it would be a combination of `$assertFullyInitialized()` with an instance variable per subclass, I suppose.

**Comment by gavinking** [comment]
Reopening because a new idea has occurred to me.

**Comment by gavinking** [comment]
So what if we require that:

- any class which leaks a self-reference in an initializer must be explicitly annotated `late`
- any constructor which leaks a self-reference must be explicitly annotated `late`
- any class which `extend`s a class or constructor annotated `late` must also be annotated `late`
- any constructor which `extend`s a class or constructor annotated `late` must also be annotated `late`

Then, any class annotated `late` must have code generated to detect access to an incompletely-initialized instance, probably in its getters. Indeed, if we just treated all its references as `late` when generating code, that might be sufficient.

Unfortunately this would be a breaking change to code that used to pass the typechecker in 1.1. But it would at least fix the unsoundness, I believe.

**Comment by chochos** [comment]
Wouldn't it just be easier to add a runtime check to any getters/attributes/methods that access any `late` members? In this case, `x` would have a runtime check similar to the one we add to `self`, and since it's self-referencing it would throw at runtime.

**Comment by chochos** [comment]
That way something like `shared Integer hash=1` that doesn't involve `self` (or `x`) wouldn't need any runtime checks, it's just a normal attribute.

**Comment by gavinking** [comment]
No that's not the problem. There is already a runtime check every time you access a late member.

**Comment by chochos** [comment]
Yes but (at least in this particular case) adding that same check to any attributes that refer to late attributes would fix it, wouldn't it?

**Comment by RossTate** [comment]
I think my proposal in #1156 incorporates a lot of your ideas above, @gavinking.

**Comment by gavinking** [comment]
Unfortunately w're not going to have time to address this in Ceylon 1.2. Moving to 1.3
